src/Features/Core/Portable/Diagnostics/DiagnosticResultSerializer.cs

```diff
@@ -122,46 +120,46 @@ internal static class DiagnosticResultSerializer
             var count = 0;
 
             writer.WriteInt32(diagnostics.Count);
-            foreach (var kv in diagnostics)
+            foreach (var (documentId, data) in diagnostics)
             {
-                kv.Key.WriteTo(writer);
-                serializer.WriteTo(writer, kv.Value, cancellationToken);
+                documentId.WriteTo(writer);
+                serializer.WriteDiagnosticData(writer, data, cancellationToken);
 
-                count += kv.Value.Length;
+                count += data.Length;
             }
 
             return count;
         }
 
-        private static ImmutableDictionary<DocumentId, ImmutableArray<DiagnosticData>> Deserialize(
+        private static ImmutableDictionary<DocumentId, ImmutableArray<DiagnosticData>> ReadDiagnosticDataMap(
             ObjectReader reader,
             DiagnosticDataSerializer serializer,
             Project project,
             CancellationToken cancellationToken)
         {
             var count = reader.ReadInt32();
+
             var map = ImmutableDictionary.CreateBuilder<DocumentId, ImmutableArray<DiagnosticData>>();
             for (var i = 0; i < count; i++)
             {
                 var documentId = DocumentId.ReadFrom(reader);
                 var document = project.GetDocument(documentId);
 
-                var diagnostics = serializer.ReadFrom(reader, document, cancellationToken);
+                var diagnostics = serializer.ReadDiagnosticData(reader, project, document, cancellationToken);
 
-                if (document?.SupportsDiagnostics() == false)
+                // drop diagnostics for non-null document that doesn't support diagnostics
+                if (diagnostics.IsEmpty || document?.SupportsDiagnostics() == false)
```

Comment:
💭 Is it possible for the data to be written to a cache in such a way that documents without data simply use the last known data? If so, it may be possible to end up in a state where diagnostics for a document don't get cleared.

Reply:
I see, having an entry in the dictionary with empty list of diagnostics isn't the same as not having the entry at all. I'll preserve that distinction.

Previously `ReadDiagnosticData` may have returned `null` (`StrongBox<ImmutableArray<DiagnosticData>>`) and `GetOrDefault(diagnostics)` was added to the map, which resulted in `default(ImmutableArray<DiagnosticData>)` added to the map, which the code that used the map did not expect. This happened only when some version mismatched or on an unexpected exception.